src/plugins/csvstorage/csvstorage.c

```diff
@@ -0,0 +1,350 @@
+	while((cur = ksNext(ks)) != NULL)
+	{
+		if(!strcmp(keyString(keyGetMeta(cur, "csvOrder")), itostr(buf, n, sizeof(buf)-1)))
+			return cur; 
+	}
+	return NULL;
+}
+int elektraCsvstorageGet(Plugin *handle, KeySet *returned, Key *parentKey)
+{
+	if (!strcmp(keyName(parentKey), "system/elektra/modules/csvstorage"))
+	{
+		KeySet *contract = ksNew (30,
+				keyNew ("system/elektra/modules/csvstorage",
+					KEY_VALUE, "csvstorage plugin waits for your orders", KEY_END),
+				keyNew ("system/elektra/modules/csvstorage/exports", KEY_END),
+				keyNew ("system/elektra/modules/csvstorage/exports/get",
+					KEY_FUNC, elektraCsvstorageGet, KEY_END),
+				keyNew ("system/elektra/modules/csvstorage/exports/set",
+					KEY_FUNC, elektraCsvstorageSet, KEY_END),
+#include ELEKTRA_README(csvstorage)
+				keyNew ("system/elektra/modules/csvstorage/infos/version",
+					KEY_VALUE, PLUGINVERSION, KEY_END),
+				KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		return 1; /* success */
+	}
+	/* get all keys */
+	KeySet *config = elektraPluginGetConfig(handle);
+	Key *delimKey = ksLookupByName(config, "/delimiter", 0);
+	Key *readHeaderKey = ksLookupByName(config, "/useheader", 0);
+	Key *key;
+	KeySet *header = ksNew(0, KS_END);
+	Key *dirKey;
+	Key *cur;
+	FILE *fp = NULL;
+	char *col;
+	char buf[15];
+	int nr_keys = 1;
+	
+	char delim = ';';
+	if(delimKey)
+		delim = ((char *)keyString(delimKey))[0];
+	
+	uint8_t useHeader = 0;
+	if(readHeaderKey)
+	{
+		if((((char *)keyString(readHeaderKey))[0] - '0') == 1)
+			useHeader = 1;
+	}
+	const char *fileName;
+	fileName = keyString(parentKey);
+	fp = fopen(fileName, "rb");
+	if(!fp)
+	{
+		ELEKTRA_SET_ERROR(116, parentKey, "couldn't open file");
+		return -1;
+	}
+
+	unsigned long length = 0;
+	length = getLineLength(fp);
+	if(length == 0)
+	{
+		ELEKTRA_ADD_WARNING(118, parentKey, "Empty file");
+		fclose(fp);
+		return 1;
+	}
+
+	char *lineBuffer;
+	lineBuffer = malloc(length * sizeof(char)+1);
+	if(!lineBuffer)
+	{
+		ELEKTRA_SET_ERROR(119, parentKey, "Out of memory");
+		return -1;
+	}
+	if(!fgets(lineBuffer, length, fp))
+	{
+		ELEKTRA_SET_ERROR(116, parentKey, "Cant read File");
+		return -1;
+	}
+	
+	unsigned long columns = 0;
+	columns = getColumnCount(lineBuffer, delim);
+
+	unsigned long colCounter = 0;
+	unsigned long lineCounter = 0;
+	unsigned long offset = 0;
+	if(useHeader == 1)
+	{
+		colCounter = 0;
+		offset = 0;
+		while((col = parseLine(lineBuffer, delim, offset)) != NULL)
+		{
+			offset += strlen(col)+1;
+			key = keyDup(parentKey);
+			keyAddBaseName(key, col);
+			keySetMeta(key, "csvOrder", itostr(buf, colCounter, sizeof(buf)-1));
+			ksAppendKey(header, key);
+			++colCounter;
+		}
+	}
+	else
+	{
+		colCounter = 0;
+		while(colCounter < columns)
+		{
+			key = keyDup(parentKey);
+			col = itostr(buf, colCounter, sizeof(buf)-1);
+			keyAddBaseName(key, col);
+			keySetMeta(key, "csvOrder", col);
+			ksAppendKey(header, key);
+			++colCounter;
+		}
+	}
+
+	while(!feof(fp))
```

Comment:
the first loop fills the keyset that is used for naming the keys with the values of the header line, the second loop just creates names 1..columns when no header is used.

gonna add some comments to all files later.